app/models/miq_smart_proxy_worker/runner.rb

```diff
@@ -56,7 +52,7 @@ def heartbeat_thread
 
   def do_work
     if @tid.nil? || !@tid.alive?
-      unless @tid.try(:status)
+      if @tid.try(:status)
```

Comment:
Figure out what's going on here.

Reply:
So if it is !alive? (aka status is running or sleeping) then status would be "aborting", false, or nil according to https://ruby-doc.org/core-2.5.0/Thread.html#method-i-status so `unless @tid.status` would have hit on only `nil` or `false` so if it were `"aborting"` it wouldn't go in this block.

I don't know if that was intentional, but I believe that is what it was doing :smile:

Reply:
Yeah, I put this comment here to track it as a TODO item for me. I was going to try to figure out what the logic was doing and make it do something better haha.

Reply:
It looks like the intention was to create the thread if it hasn't been created or if it is dead.

Reply:
I think they might have not wanted to join the thread if it was still aborting, since status of false or nil indicate the thread is exited. We could just pass a timeout duration to join and kill the thread if it doesn't join cleanly.